keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+to deallocate resources.
+
+#### Post-filter
+
+This is called when no suitable node could be found. If the Pod depends on ResourceClaims with delayed
+allocation, then deallocating one or more of these ResourceClaims may make the
+Pod schedulable after allocating the resource elsewhere. Therefore each
+ResourceClaim with delayed allocation is checked whether all of the following
+conditions apply:
+- allocated
+- not currently in use
+- it was the reason why some node could not fit the Pod, as recorded earlier in
+  Filter
+
+One of the ResourceClaims satisfying these criteria is picked randomly and its
+Phase is set to Deallocate. This may make it possible to run the Pod
+elsewhere. If it still doesn't help, deallocation may continue with another
+ResourceClaim, if there is one. To prevent deallocating all resources merely
+because the scheduler happens to check quickly, the next deallocation will only
+requested when there is none currently running.
+
+At the moment, the scheduler has no information that might enable it to
+prioritize which resource to deallocate first. Future extensions of this KEP
+might attempt to improve this.
+
+#### Pre-score
+
+This is passed a list of nodes that have passed filtering by the resource
+plugin and the other plugins. The PotentialNodes field of unallocated
+ResourceClaims with delayed allocation gets updated now if the field doesn't
+match the current list already.
+
+#### Reserve
+
+A node has been chosen for the Pod.
+
+If using delayed allocation and the resource has not been allocated yet,
+the SelectedNode field of the ResourceClaim
+gets set here and the scheduling attempt gets stopped for now. It will be
+retried when the ResourceClaim status changes.
+
+If all resources have been allocated already,
+the scheduler adds the Pod to the ReservedFor field of its ResourceClaims to ensure that
+no-one else gets to use those.
+
+If some resources are not allocated yet or reserving an allocated resource
+fails, the scheduling attempt needs to be aborted and retried at a later time
+or when the ResourceClaims change.
+
+#### Unreserve
+
+The scheduler removes the Pod from the ReservedFor field because it cannot be scheduled after
+all.
+
+### kubelet
+
+#### Managing resources
+
+kubelet must ensure that resources are ready for use on the node before running
+the first Pod that uses a specific resource instance and make the resource
+available elsewhere again when the last Pod has terminated that uses it. For
+both operations, kubelet calls a resource kubelet plugin as explained in the next
+section.
+
+Pods that are not listed in ReservedFor or where the ResourceClaim doesn't
+exist at all must not be allowed to run. Instead, a suitable event must be
+emitted which explains the problem. Such a situation can occur as part of
+downgrade scenarios.
+
+In addition, kubelet must remove a Pod from ResourceClaim.ReservedFor before
+deleting the Pod. If this was the last Pod on the node that uses the specific
```

Comment:
How does the kubelet know its the last pod and its not a resource that was pre-allocated?

I am trying to imagine what the reconcile loop for this looks like on kubelet start-up.

Reply:
I'm assuming kubelet knows which Pods are currently running or meant to run. That determines whether its time to prepare or unprepare a resource.

Reply:
the meant to run part is what confused me.  i thought i could make a claim independent of a pod ever following?

Reply:
A claim that is completely independent of a pod can be allocated by using "immediate" allocation.

kubelet only gets involved when a Pod references a claim and the Pod has been scheduled onto a node.

Reply:
> I think Derek's point is that Kubelet can't easily tell when the lifetime of a resource is SUPPOSED to end. 

It doesn't need to know that because it doesn't need to do anything special. In particular it never deletes a ResourceClaim.

Let's consider the ephemeral claim case (= lifecycle is coupled). The ResourceClaim gets created when the pod gets created, with the pod as owner.

The usual allocation and scheduling happens and the pod gets to run on a node. When kubelet cleans up because a pod has a DeletionTimestamp, it makes sure that the pod is no longer listed in `ReservedFor`. For an ephemeral resource that pod will typically be the only user, so the field is now empty and the claim is considered unused.

When kubelet clears the GracePeriod, the pod gets deleted. Now the normal garbage collector kicks in and also deletes the ephemeral claim because of its owner reference.

To the resource driver that looks like normal deletion of a claim and because it's not in use, it will get deallocated and removed. Kubelet does not need to wait for this.

This is exactly the same approach as for [generic ephemeral inline volumes](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1698-generic-ephemeral-volumes). Do I need to replicate the design description of that in this KEP or is it sufficient if I summarize here?

Currently this KEP only states:
> When a Pod gets created, such a template
will be used to create a normal ResourceClaim with the Pod as owner, and then the normal allocation of the resource takes place.